WebAuthn docs update (#25922)
mdn/content: files/en-us/web/api/web_authentication_api/authenticator_data/index.md

    - Bit 4, Backup State (BS): If set, the public key credential source is currently backed up (see Bit 3 for context).
    - Bit 6, Attested Credential Data (AT): If set, the attested credential data will immediately follow the first 37 bytes of this `authenticatorData`.
    - Bit 7, Extension Data (ED): If set, extension data is present. Extension data will follow attested credential data if it is present, or will immediately follow the first 37 bytes of the `authenticatorData` if no attested credential data is present.

- **signCount** (4 bytes)
  - : A signature counter, if supported by the authenticator (set to 0 otherwise). Servers may optionally use this counter to detect authenticator cloning.
- **attestedCredentialData** (variable length)

  - : The credential that was created. This is only present during a {{domxref("CredentialsContainer.create", "navigator.credentials.create()")}} call. This is a sequence of bytes with the following format:

>     - **AAGUID** (16 bytes): The Authenticator Attestation Globally Unique Identifier, a unique number that identifies the model of the authenticator (not the specific instance of the authenticator). A relying party can use this to find out the characteristics of the authenticator by looking up its metadata statement via the [FIDO metadata service](https://fidoalliance.org/metadata/). This is only really relevant to sophisticated enterprise relying parties.

Author: I like the examples, and I like the more positive spin. I've updated it to 

> This is relevant in certain situations such as enterprise deployments or where regulatory requirements dictate a certain type of authenticator be used; it should be ignored otherwise.

    - **_credentialIdLength_** (2 bytes): The length of the credential ID that immediately follows these bytes.
    - **_credentialId_** (variable length): A unique identifier for this credential so that it can be requested for future authentications. The credential is "_credentialIdLength_" bytes long.

>     - **credentialPublicKey** (variable length): A [COSE](https://datatracker.ietf.org/doc/html/rfc8152)-encoded public key. This public key will be stored on the server associated with a user's account and be used for future authentications. If you don't want to go through the trouble of parsing COSE-ended data, you can retrieve the public key as DER-encoded `SubjectPublicKeyInfo` via {{domxref("AuthenticatorAttestationResponse.getPublicKey()")}}.

Reviewer: ```suggestion
    - **credentialPublicKey** (variable length): A [COSE](https://datatracker.ietf.org/doc/html/rfc8152)-encoded public key. This public key will be stored on the server associated with a user's account and be used for future authentications. Relying Parties can easily pull out the DER-encoded form of this public key, without parsing the COSE-encoded authenticator data, via the {{domxref("AuthenticatorAttestationResponse.getPublicKey()")}} method.
```

Author: Yup, sounds good; I've updated it.

- **extensions** (variable length)

  - : An optional [CBOR](https://datatracker.ietf.org/doc/html/rfc7049) map containing the response outputs from extensions processed by the authenticator

>     Extensions are optional and different browsers may recognize different extensions. Processing extensions is always optional for the client: if a browser does not recognize a given extension, it will just ignore it. For information on using extensions, and which ones are supported by which browsers, see [Web Authentication extensions](/en-US/docs/Web/API/Web_Authentication_API/WebAuthn_extensions).

Author: Ah, yes. I've tended to use user agent or browser everywhere for this, but I slipped up here. I'll update this one, and look out for it in other places.

Give me a ping if you notice any others.

---
title: Authenticator data
slug: Web/API/Web_Authentication_API/Authenticator_data
page-type: guide
---

{{DefaultAPISidebar("Web Authentication API")}}

> The authenticator data structure contains information from the authenticator about the processing of a credential creation or authentication request — such as the Relying Party ID Hash (rpIdHash), a signature counter, test of user presence, user verification flags, and any extensions processed by the authenticator. This page explains what is contained in the data structure.

Reviewer 2: ```suggestion
The authenticator data structure contains information from the authenticator about the processing of a credential creation or authentication request — such as the Relying Party ID Hash (`rpIdHash`), a signature counter, test of user presence, user verification flags, and any extensions processed by the authenticator. This page explains what is contained in the data structure.
```

- Via the {{domxref("AuthenticatorAttestationResponse.getAuthenticatorData", "PublicKeyCredential.response.getAuthenticatorData()")}} method made available on the {{domxref("PublicKeyCredential")}} returned from a successful {{domxref("CredentialsContainer.get", "navigator.credentials.get()")}} (authentication) call.

## Data structure

An authenticator data {{jsxref("ArrayBuffer")}} is at least 37 bytes in length, and contains the following fields:

- **rpIdHash** (32 bytes)
  - : The SHA-256 hash of the [Relying Party ID](https://w3c.github.io/webauthn/#relying-party-identifier) that the credential is scoped to. The server will ensure that this hash matches the SHA256 hash of its own relying party ID in order to prevent phishing or other man-in-the-middle attacks.
- **flags** (1 bytes)

  - : A bitfield that indicates various attributes that were asserted by the authenticator. The bits are as follows, where Bit 0 is the least significant bit and all bits not specifically mentioned below are "reserved for future use":

>     - Bit 0, User Presence (UP): If set (i.e. to `1`), the authenticator validated that the user was present through some Test of User Presence (TUP), such as touching a button on the authenticator.

Reviewer 2: ```suggestion
    - Bit 0, User Presence (UP): If set (i.e., to `1`), the authenticator validated that the user was present through some Test of User Presence (TUP), such as touching a button on the authenticator.
```
